boston-housing/data.js

```diff
@@ -15,78 +15,108 @@
  * =============================================================================
  */
 
-import * as tf from '@tensorflow/tfjs';
-import * as utils from './utils';
+const Papa = require('papaparse');
+
 
 // Boston Housing data constants:
-const TRAIN_DATA = 'train-data';
-const TRAIN_TARGET = 'train-target';
-const TEST_DATA = 'test-data';
-const TEST_TARGET = 'test-target';
+const BASE_URL =
+    'https://storage.googleapis.com/tfjs-examples/multivariate-linear-regression/data/';
+
+const TRAIN_FEATURES_FN = 'train-data.csv';
+const TRAIN_TARGET_FN = 'train-target.csv';
+const TEST_FEATURES_FN = 'test-data.csv';
+const TEST_TARGET_FN = 'test-target.csv';
+
+/**
+ *
+ * @param {Array<Object>} data Downloaded data.
+ *
+ * @returns {Promise.Array<number[]>} Resolves to data with values parsed as floats.
+ */
+const parseCsv = async (data) => {
+  return new Promise(resolve => {
+    data = data.map((row) => {
+      return Object.keys(row).map(key => parseFloat(row[key]));
+    });
+    resolve(data);
+  });
+};
+
+/**
+ * Downloads and returns the csv.
+ *
+ * @param {string} filename Name of file to be loaded.
+ *
+ * @returns {Promise.Array<number[]>} Resolves to parsed csv data.
+ */
+export const loadCsv = async (filename) => {
+  return new Promise(resolve => {
+    const url = `${BASE_URL}${filename}`;
+
+    console.log(`  * Downloading data from: ${url}`);
+    Papa.parse(url, {
+      download: true,
+      header: true,
+      complete: (results) => {
+        resolve(parseCsv(results['data']));
+      }
+    })
+  });
+};
+
+
 
 /** Helper class to handle loading training and test data. */
 export class BostonHousingDataset {
   constructor() {
-    this.dataset = null;
+    // Arrays to hold the data.
+    this.trainFeatures = null;
+    this.trainTarget = null;
+    this.testFeatures = null;
+    this.testTarget = null;
+    // Metadata
     this.trainSize = 0;
     this.testSize = 0;
-    this.trainBatchIndex = 0;
-    this.testBatchIndex = 0;
-
-    this.NUM_FEATURES = 12;
   }
 
   get numFeatures() {
-    return this.NUM_FEATURES;
+    return this.trainFeatures[0].length;
```

Comment:
Now that we are using `this.trainFeatures[0]` length as the number of features, I feel this will throw an error if `numFeatures` is accessed before calling `loadData`. We should either call `loadData` in the constructor or check if it is null before accessing the element.

Reply:
Agreed, updated the object to keep track of its load state in a private boolean member variable and to throw an error if the features are accessed before the data is loaded.

I've also removed the other two pieces of metadata, that weren't actually being used anymore (since tensor2d is able to determine the shape automatically from input type number[][]).